systemlib cherrypicks for r2.4

perfinion opened the pull request:
This cherry-picks a few fixes into r2.4 to make systemlib builds work.

This cherry-picks from:
https://github.com/tensorflow/tensorflow/pull/44144
https://github.com/tensorflow/tensorflow/pull/44222
https://github.com/tensorflow/tensorflow/pull/44488

The first two are already merged into master and the last is a tiny fix for the merge conflict which was squashed into the protobuf commit

perfinion committed:
systemlibs: Update to build against system GRPC

- Add libgpr
    Newer grpc-1.28 has a libgpr.so that is also needed during link time
    so add it to the linkopts
- Add starlark files
    Several starlark files are load()'d from the GRPC repo, vendor them
    or add stubs as appropriate when using the system version of grpc.
- grpc WORKSPACE deps
    Several deps were loaded in WORKSPACE that were needed by grpc, they
    are not needed when building against the system but are difficult to
    stub out causing the build to fail.
    grpc_extra_deps.bzl is provided to load all the requirements, so use
    that from WORKSPACE instead of directly loading each individually.
    This is also more maintainable going forward since there is less to
    keep in sync in TF's WORKSPACE file.

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
systemlibs: unbundle @dill_archive dep

Signed-off-by: Jason Zaman <jason@perfinion.com>

perfinion committed:
systemlibs: unbundle @tblib_archive dep

Signed-off-by: Jason Zaman <jason@perfinion.com>

Flamefire committed:
Use upstream WELL_KNOWN_PROTO_MAP to include dependencies

Flamefire committed:
Introduce PROTOBUF_INCLUDE_PATH

This avoids having to install all dependencies into the same folder
Default set so existing behavior is not changed

perfinion commented:
@Flamefire Can you add the consent to cherry-pick the first two commits into r2.4
thanks!

Flamefire commented:
@googlebot I consent.

@perfinion Why are the headers required for protobuf? Haven't been required on our system and having them (actually the copy of them) caused problems if the protobuf version didn't fully match and headers were missing or added

perfinion commented:
> @perfinion Why are the headers required for protobuf? Haven't been required on our system and having them (actually the copy of them) caused problems if the protobuf version didn't fully match and headers were missing or added

@Flamefire I get errors like this if i remove the `hdrs = HEADERS,` line from protobuf.BUILD

```
ERROR: /home/jason/code/tf/tensorflow/tensorflow/tools/proto_text/BUILD:31:10: undeclared inclusion(s) in rule '//tensorflow/tools/proto_text:gen_proto_text_functions':
this rule is missing dependency declarations for the following files included by 'tensorflow/tools/proto_text/gen_proto_text_functions.cc':
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/io/coded_stream.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/io/zero_copy_stream.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/io/zero_copy_stream_impl_lite.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/descriptor.pb.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/arena.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/descriptor.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/map.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/repeated_field.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/text_format.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/util/json_util.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/util/type_resolver_util.h'
  'bazel-out/k8-opt-exec-50AE0418/bin/external/com_google_protobuf/google/protobuf/compiler/importer.h'
```

my headers are in eg `/usr/include/google/protobuf/io/coded_stream.h`  so they _should_ just be automatically available because my compiler by default looks in /usr/include. Have you seen anything like this? I dont need it for any other deps it only happens for protobuf headers. I also really dislike needing to hardcode headers, the jsoncpp headers were a massive headache before you fixed them too but i'm not sure how cuz these arent a different include path :( .

Flamefire commented:
This very much looks like there are headers still there. I removed the copy here (same as for jsoncpp) so the system location should be used instead. Is it possible you are doing a "dirty" build? I.e. those headers are still there and were copied before my PR removed the copy and hence are used due to Bazel auto-adding those as include paths? If so simply deleting those headers is very much recommended as they might even be from an outdated protobuf installation.

I'd recommend resolving that first before merging

mihaimaruseac commented:
Waiting on #44522

perfinion commented:
I finally figured out the protobuf issues, the culprit was `--disk_cache=` I have enabled. Even with `bazel clean --expunge` the old stuff was still being pulled out of the cache and re-used. I unreliably managed to make it work even with keeping the cache around. Doing a clean --expunge and changing some stuff in the protobuf.BUILD file would get it to work. Strangely adding `tags = ["no-cache"],` did not always work. The most reliably way to fix it was to just nuke my --disk_cache= dir.

I dropped the protobuf commit from the top of this series (and rebased on top of the r2.4 branch). It should be okay to merge 
into r2.4 now. I also approved https://github.com/tensorflow/tensorflow/pull/44522 which should go into master

perfinion committed:
systemlibs: unbundle typing_extensions

Signed-off-by: Jason Zaman <jason@perfinion.com>
